tools/c7n_azure/tests/azure_common.py

```diff
@@ -12,115 +12,306 @@
+        myvcr.register_matcher('azure-matcher', self._azure_matcher)
+        myvcr.match_on = ['azure-matcher', 'method']
+        myvcr.register_serializer('azure-json', AzureSerializer())
+        myvcr.serializer = 'azure-json'
+        myvcr.path_transformer = AzureVCRBaseTest._json_extension
 
-        # Block recording when using fake token (tox runs)
+        # Block recording when using fake token (generally only used on build servers)
         if os.environ.get(constants.ENV_ACCESS_TOKEN) == "fake_token":
             myvcr.record_mode = 'none'
 
         return myvcr
 
-    def azure_matcher(self, r1, r2):
+    def _azure_matcher(self, r1, r2):
         """Replace all subscription ID's and ignore api-version"""
         if [k for k in set(r1.query) if k[0] != 'api-version'] != [
                 k for k in set(r2.query) if k[0] != 'api-version']:
             return False
 
-        r1_path = re.sub(
-            r"[\da-zA-Z]{8}-([\da-zA-Z]{4}-){3}[\da-zA-Z]{12}",
-            DEFAULT_SUBSCRIPTION_ID,
-            r1.path)
-        r2_path = re.sub(
-            r"[\da-zA-Z]{8}-([\da-zA-Z]{4}-){3}[\da-zA-Z]{12}",
-            DEFAULT_SUBSCRIPTION_ID,
-            r2.path)
+        r1_path = AzureVCRBaseTest._replace_subscription_id(r1.path)
+        r2_path = AzureVCRBaseTest._replace_subscription_id(r2.path)
 
         r1_path = r1_path.replace('//', '/')
         r2_path = r2_path.replace('//', '/')
 
         return r1_path == r2_path
 
-    def request_callback(self, request):
+    def _request_callback(self, request):
         """Modify requests before saving"""
-        if "/subscriptions/" in request.url:
-            request.uri = re.sub(
-                r"[\da-zA-Z]{8}-([\da-zA-Z]{4}-){3}[\da-zA-Z]{12}",
-                DEFAULT_SUBSCRIPTION_ID,
-                request.url)
+        request.uri = AzureVCRBaseTest._replace_subscription_id(request.uri)
+        request.uri = AzureVCRBaseTest._replace_tenant_id(request.uri)
+
         if request.body:
             request.body = b'mock_body'
+
+        # Request headers serve no purpose as only URI is read during a playback.
+        request.headers = None
+
         if re.match('https://login.microsoftonline.com/([^/]+)/oauth2/token', request.uri):
             return None
         if re.match('https://login.microsoftonline.com/([^/]+)/oauth2/token', request.uri):
             return None
         return request
 
+    def _response_callback(self, response):
+        if self.is_playback():
+            if 'data' in response['body']:
+                body = json.dumps(response['body']['data'])
+                response['body']['string'] = body.encode('utf-8')
+                response['headers']['content-length'] = [str(len(body))]
+
+            return response
+
+        headers = {}
+        for key in response['headers']:
+            if key.lower() not in self.FILTERED_HEADERS:
+                headers[key.lower()] = response['headers'][key]
```

Comment:
May be cleaner to do with dictionary comprehension, something like...
headers = {k.lower():v (k,v) in response['headers'].items() if k.lower() not in self.FILTERED_HEADERS}